File: graphs/redundant_connection.py
'''
LeetCode - problem 684
'''

    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        parent = [i for i in range(len(edges)+1)]
        def find(x):
            if x != parent[x]:
                parent[x] = find(parent[x])
            return parent[x]
        
        def union(x,y):
            rootx = find(x)
            rooty = find(y)
            if rootx == rooty:
                return [x,y]
            else:#connect
                parent[rooty] = rootx
                
        for x,y in edges:
            res = union(x,y)
            if res:
                return res

File: graphs/redundant_connection.cpp
//LeetCode - problem 684

	int FindParent(int node , vector<int>& parent){
		if(parent[node] == node) return node;
		return parent[node] = FindParent(parent[node],parent);
	}

	bool hascycle(int u , int v,  vector<int> &parent){
		int u_parent = FindParent(u,parent);
		int v_parent = FindParent(v,parent);
		return u_parent == v_parent;
	}
public:
	vector<int> findRedundantConnection(vector<vector<int>>& arr) {
		int n = arr.size();
		vector<int> parent(n+1);
		for(int i = 0 ; i <= n ; i++) parent[i] = i;
		int start = -1 , end = -1;
		for(int i = 0 ; i < n ; i++){
			int u = arr[i][0] , v = arr[i][1];
			if(hascycle(u,v,parent)){
				start = u, end = v;
				continue;
			}
			parent[FindParent(v,parent)] = u;
		}
		return {start,end};
	}
